storage/core/src/types.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2022-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Database backend support building blocks
+
+use core::ops::Range;
+
+/// Used to identify particular key-value map in the database
+#[derive(Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub struct DbMapId(usize);
+
+impl DbMapId {
+    /// New index
+    pub const fn new(id: usize) -> Self {
+        DbMapId(id)
+    }
+
+    /// Get the index as usize
+    pub const fn as_usize(&self) -> usize {
+        self.0
+    }
+}
+
+/// Number of key-value maps in the database
+#[derive(Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub struct DbMapCount(usize);
+
+impl DbMapCount {
+    /// Convert to `usize`
+    pub const fn as_usize(self) -> usize {
+        self.0
+    }
+
+    /// Iterator over all database indices
+    pub fn indices(self) -> impl Iterator<Item = DbMapId> + ExactSizeIterator {
+        (0..self.0).map(DbMapId::new)
+    }
+}
+
+/// Associate data of type `T` with each DB map
+///
+/// This is used to associate certain data with each DB map. The data can be metadata containing
+/// names and other information about the maps, like [DbMapDesc]. It can also be used by the
+/// backend to store for example handles to open key-value maps.
+///
+/// The interface of this type is deliberately constrained to make it hard to use inappropriately.
+/// It can be constructed in two ways:
+/// 1. The [Self::new] method takes in the number of DB maps. The map count is in a type wrapper to
+///    make it harder to use arbitrary number there.
+/// 2. The [Self::transform] method is used to derive one piece of data from another. It preserves
+///    the data count and ensures the DB map IDs remain consistent between the original and the
+///    derived [DbMapsData] instance.
+/// 3. The data can only be accessed by a [DbMapId] so it's only indexed into using a valid DB
+///    map identifier.
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct DbMapsData<T>(Vec<T>);
```

Comment:
I maintain that this should be an array, not a Vec... the size of this array is determined at compile-time... but again, you don't have to do this now.

What hurts my eyes most in this is the excessive usage of the index operator for every operation, which is constantly done with a boundary checks. The reason this isn't so bad is that it's accompanied with I/O operations, so in comparison it's negligible. But from a type system point of view, this type's size should be fixed at compile-time.

Reply:
Well I see the point but the tradeoff is that the compiler will generate the backend for each map count with almost identical code which seems a bit excessive. The backends are deliberately independent of the schema, so you can build them once and apply to many databases. It also provides conceptual separation at design level where the schema does not leak into type of the backend.

Reply:
By the way changing this to an array does not eliminate the boundary checks

Reply:
Out of curiosity, isn't the separation you're referring to achieved by having `T` as a generic type?

Also I think the checks will be eliminated because all accesses are done through enums, which are defined at compile-time. The compiler is smart enough to do that.

Reply:
> Out of curiosity, isn't the separation you're referring to achieved by having `T` as a generic type?

I don't see how that's related. The `T` refers to data stored for each db map and can be picked by the backend at will. Nothing to do with schema or number of maps as such. Maybe I misunderstand your comment.
 
> Also I think the checks will be eliminated because all accesses are done through enums, which are defined at compile-time. The compiler is smart enough to do that.

Currently, there are no enums as far as I can tell. Maybe you have a different design in mind. At the front end, the db map is identified by a unit struct, not an enum. If we were to change the db map index to be represented by an enum, then in order for the compiler to take the advantage of it to eliminate bounds checking, the backend would have to be parametrized by the enum.

Reply:
I meant the enum at the application layer.

Not important. Maybe one day I'll try it and then I'll figure out what you're saying myself 😄